File: desktop/src/lib/bridge/files.ts
/**
 * Plan 2b.2.c.3 — frontend wrappers for the Rust Tauri commands defined in
 * desktop/src-tauri/src/commands/files.rs. Replaces the no-op stubs in
 * peripheral-stubs.ts; consumers (InlineImage in AgentMessages, sdk-message-renderer,
 * tool-activity-item) retarget to import from here.
 */
import { commands } from './generated'

/**
 * Read a file from disk and return its base64-encoded bytes.
 * Returns null if the file is missing or unreadable.
 */
export async function readAttachment(localPath: string): Promise<string | null> {
  try {
    const result = await commands.readAttachment(localPath)
    return result.status === 'ok' ? result.data : null
  } catch {
    return null
  }
}

export interface SaveImageArgs {
  localPath: string
  filename: string
  mediaType: string
}

/**
 * Open the native save-as dialog and copy `localPath` to the chosen target.
 * Returns true on success, false on user cancel or copy failure.
 */
export async function saveImageAs(args: SaveImageArgs): Promise<boolean> {
  try {
    const result = await commands.saveImageAs({
      local_path: args.localPath,
      filename: args.filename,
      media_type: args.mediaType,
    })
    return result.status === 'ok' ? result.data : false
  } catch {
    return false
  }
}

File: desktop/src-tauri/src/commands/files.rs
//! Tauri commands for file attachments.
//!
//! Plan 2b.2.c.3 — replaces the frontend `peripheral-stubs.readAttachment` /
//! `peripheral-stubs.saveImageAs` no-ops. Consumers: `InlineImage` in
//! AgentMessages and `screenshot-result.tsx`.

use base64::Engine;
use serde::Deserialize;
use std::path::PathBuf;
use tauri::AppHandle;
use tauri_plugin_dialog::DialogExt;

/// Read a file from disk and return its bytes as base64.
/// Used by InlineImage / screenshot-result to render local attachments
/// without a `file://` blob URL (which is blocked by Tauri's CSP).
#[tauri::command]
#[specta::specta]
pub async fn read_attachment(path: String) -> Result<String, String> {
    let bytes = tokio::fs::read(&path)
        .await
        .map_err(|e| format!("read {path}: {e}"))?;
    Ok(base64::engine::general_purpose::STANDARD.encode(&bytes))
}

#[derive(Debug, Deserialize, specta::Type)]
pub struct SaveImageArgs {
    pub local_path: String,
    pub filename: String,
    /// MIME type forwarded from the frontend; reserved for future use
    /// (e.g. choosing the save-dialog filter dynamically).
    #[allow(dead_code)]
    pub media_type: String,
}

/// Open a save-as dialog for the user, then copy `local_path` to the chosen
/// destination. Returns `true` if the user picked a destination + copy
/// succeeded; `false` if the user cancelled.
#[tauri::command]
#[specta::specta]
pub async fn save_image_as(app: AppHandle, args: SaveImageArgs) -> Result<bool, String> {
    let dest: Option<PathBuf> = app
        .dialog()
        .file()
        .add_filter("Images", &["png", "jpg", "jpeg", "webp"])
        .set_file_name(&args.filename)
        .blocking_save_file()
        .and_then(|fp| fp.into_path().ok());

    let Some(dest) = dest else {
        return Ok(false);
    };

    tokio::fs::copy(&args.local_path, &dest)
        .await
        .map_err(|e| format!("copy {} -> {}: {e}", args.local_path, dest.display()))?;
    Ok(true)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn read_attachment_round_trips_a_small_file() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("hello.txt");
        tokio::fs::write(&path, b"hello").await.unwrap();
        let b64 = read_attachment(path.to_string_lossy().into_owned())
            .await
            .unwrap();
        // 'hello' base64-encoded is 'aGVsbG8='
        assert_eq!(b64, "aGVsbG8=");
    }

    #[tokio::test]
    async fn read_attachment_returns_error_for_missing_file() {
        let res = read_attachment("/definitely/does/not/exist".to_string()).await;
        assert!(res.is_err());
    }
}
